lib/fake_async.dart

```
@@ -211,39 +211,58 @@ class FakeAsync {
+          break;
+        }
       }
 
-      if (flushPeriodicTimers) return _timers.isNotEmpty;
+      if (!runNextTimer(until: absoluteTimeout)) {
+        if (_timers.isEmpty) break;
 
-      // Continue firing timers until the only ones left are periodic *and*
-      // every periodic timer has had a change to run against the final
-      // value of [_elapsed].
-      return _timers
-          .any((timer) => !timer.isPeriodic || timer._nextCall <= _elapsed);
-    });
+        // TODO(nweiz): Make this a [TimeoutException].
+        throw StateError('Exceeded timeout $timeout while flushing timers');
+      }
+    }
   }
 
-  /// Invoke the callback for each timer until [predicate] returns `false` for
-  /// the next timer that would be fired.
+  /// Elapses time to run one timer, if any timer exists.
+  ///
+  /// Running one timer at a time, rather than just advancing time such as
+  /// with [elapse] or [flushTimers], can be useful if a test wants to run
+  /// its own logic between each timer event, for example to verify invariants
+  /// or to end the test early in case of an error.
   ///
-  /// Microtasks are flushed before and after each timer is fired. Before each
-  /// timer fires, [_elapsed] is updated to the appropriate duration.
-  void _fireTimersWhile(bool Function(FakeTimer timer) predicate) {
+  /// Microtasks are flushed before identifying the timer to run,
+  /// and again after the timer runs.
+  /// Before the timer runs, [elapsed] is updated to the appropriate value.
+  ///
+  /// When [until] is non-null, only timers due up until the given time
+  /// will be considered, in terms of [elapsed].
+  ///
+  /// Because multiple timers may be due at the same time, a call to this
+  /// method may leave the time advanced to where other timers are due.
+  /// Doing an `elapse(Duration.zero)` afterwards may trigger more timers.
+  ///
+  /// Returns `true` if a timer was run, `false` otherwise.
+  bool runNextTimer({Duration? until}) {
```

Is this code re-entrancy safe? (Likely not, code rarely is if it's not designed for it.)

That is, if someone calls `runNextTimer` from a microtask or timer event callback, so it's *inside a running event loop*, will something break?
(Should we check for it and maybe throw a state error?)

---

Hmm, good question.

In fact, when I stared at this code today to try to pin down an answer, I discovered a bug in the existing code:
* #88
* #89

---

It's actually normal and useful, though, that `elapse` gets called from microtasks within `flushTimers` (and so that `runNextTimer` gets called from microtasks within `runNextTimer`).

Specifically, we write tests like
```dart
  test('thing gets handled', () => awaitFakeAsync((async) async {
    await scheduleThing(delay: const Duration(seconds: 1));
    async.elapse(const Duration(seconds: 1));
    checkThingWasHandled();
  }));
```

where [`awaitFakeAsync`](https://github.com/zulip/zulip-flutter/blob/e85d6c5fbc5b5df528c857ba5e23fab7d044296d/test/fake_async.dart) is the helper described in #84. So that `async.elapse` call comes from a microtask, fired by the `flushTimers` call in the implementation of `awaitFakeAsync` (with current `fake_async`), or the `runNextTimer` call in my draft revision of `awaitFakeAsync` (using this PR).

---

Fortunately I think `runNextTimer`'s internal assumptions are perfectly safe with re-entrancy:
* After the first `flushMicrotasks` call, the only new fact it can assume is that the microtask queue is empty. And `flushMicrotasks` ensures that before it returns, regardless of what the microtasks do. So if a microtask calls `runNextTimer`, things are no different from just calling `runNextTimer` twice in succession.
* Conversely, on entering the `flushMicrotasks` call at the end, the only post-condition that remains for `runNextTimer` to establish is that the microtask queue is empty (as that `flushMicrotasks` call is the very last thing it does). If a microtask calls `runNextTimer`, that call will do its own `flushMicrotasks`, so it's just as if it had happened after the outer call finished.
* Finally, the same thing is true of when the timer callback gets fired, with that `timer._fire()` call just before `flushMicrotasks`…
  * … except that when the timer is periodic, `FakeTimer._fire` updates `_nextCall` only *after* calling the timer callback. That sounds like a re-entrancy bug, and indeed it is: #88. Starting to write this paragraph is what led me to discover that, and then I took a detour to confirm that it is and write up a PR to fix it (#89).

Beyond `runNextTimer` itself, this library's two event loops that call it (in `elapse` and `flushTimers`) are also safe with re-entrancy, by the same reasoning that I give in #89 to argue that that fix is complete.

---

The possibility of re-entrancy does call for a bit of care, though, when looking at a call site of `runNextTimer` and thinking about its behavior. Here's a paragraph I'll add to the doc:
```dart
  /// If microtasks or timer callbacks make their own calls to methods on this
  /// [FakeAsync], then a call to this method may indirectly cause more timers
  /// to run beyond the timer it runs directly, and may cause [elapsed] to
  /// advance beyond [until].  Any such timers are ignored in the return value.
```

In terms of the use cases I put in the doc:
* to verify invariants between each timer event: For this use case, one would avoid calling `runNextTimer` except at the call site that will verify the invariants (and would avoid calling `elapse` and `flushTimers` entirely).
* to end the test early in case of an error (which is the use case I actually have): For this use case I think this wrinkle is perfectly fine. If the test body (e.g., the callback passed to `awaitFakeAsync`) has already completed with an error, then it won't be making any more `FakeAsync` method calls; and only the test should be seeing the `FakeAsync` instance in the first place.